### ISSUE TITLE
Can we have the report in Excel?

### ISSUE THREAD

**Issue by lchong0412** [TestStack/TestStack.BDDfy#206]
Is it possible to generate test report in Excel? If it is, how do we do that?

Thanks

**Comment by mwhelan** [comment]
Yes, it is. The best way is to create a custom batch processor. You could follow the example in the [docs](http://bddfy.teststack.net/v1.0/docs/customizing-reports). The batch processor will run after all of the tests have completed and will give you a list of stories with all of the results for the scenarios/steps. You would just need to loop through these and write them out to Excel.

Let us know if you need any more info.

**Comment by lchong0412** [comment]
Thanks, I will give it a try!

**Comment by lchong0412** [comment]
Sorry, another question, is it possible to access the seleno host in the processor class so that I can take a screenshot if a test fails?

**Comment by MehdiK** [comment]
What's the use case? Seleno [takes screenshot on failures](https://github.com/TestStack/TestStack.Seleno/blob/master/src/TestStack.Seleno/Configuration/Interceptors/CameraProxyInterceptor.cs#L6) automatically.

**Comment by MehdiK** [comment]
Also, no need to be sorry. There are unfortunately not much documentation for Seleno (it's in the backlog though). We're here to answer questions anytime.

**Comment by lchong0412** [comment]
Cool, that's awesome

**Comment by lchong0412** [comment]
I tried it by intentially failing an assertion, but I couldn't find the screenshot anywhere, do I need to configure something in the SelenoHost?

**Comment by MehdiK** [comment]
@robdmoore could you shed some light on this please? Thanks.

**Comment by MehdiK** [comment]
@lchong0412 What do you mean by 'get the current running Scenario'?

**Comment by robdmoore** [comment]
Yep, you need to configure a directory to drop the screenshots into, like this: https://github.com/TestStack/TestStack.Seleno/blob/master/src/TestStack.Seleno.AcceptanceTests/Configuration/ScreenshotTests.cs#L26

**Comment by lchong0412** [comment]
@robdmoore Hi, I did this but as I stated before, the problem was the assertion failure doesn't seem to be handled by the Interceptor, it's not taking any screenshot when I just put Assert.True(1==2)

**Comment by lchong0412** [comment]
@MehdiK All I want to do is in my Dispose (teardown) method of a test class, I can check whether my current test method (scenario) is passed or failed, so that if it fails, I can take a screenshot. I know I can handle the test result in the Processor pipeline, but then I won't be able to access the SelenoHost to take the photo in there

**Comment by robdmoore** [comment]
Yep, so Seleno can only instigate a screenshot automatically from within Seleno code e.g. You try and get an element that doesn't exist.

If you want to do s screenshot for assertion errors you need to do your own try/catch and call into SelenoHost to take the screenshot.

**Comment by robdmoore** [comment]
Try putting the try/catch around the this.Bddfy call. SelenoHost should be statically available if you follow the patterns in Seleno.

**Comment by lchong0412** [comment]
@robdmoore thanks for your reply, that explains. But I'm using fluent style, so it means i will have to try catch every test method?

**Comment by robdmoore** [comment]
Yes.

Perhaps you can use AOP e.g. PostSharp or Fody.MrAdvice or there might be a way to do this with BDDfy (@mehdik?)

**Comment by mwhelan** [comment]
How about this? Code review please, it's knocked up pretty quickly :-)

Just call .BDDfyX at the end of your fluent chain instead of the normal .BDDfy.

```
public static Story BDDfyX(this object testObject, string scenarioTitle = null)
{
    try
    {
        return testObject.BDDfy(scenarioTitle);
    }
    catch (Exception exception)
    {
        Host.Instance.Application.Camera.TakeScreenshot();
    }
    return null;
}
```

**Comment by mwhelan** [comment]
Alternatively, you might consider writing a custom exception processor. I'd probably stick with the extension method....

```
public class SelenoExceptionProcessor : IProcessor
{
    public ProcessType ProcessType { get {return ProcessType.ProcessExceptions;} }

    public void Process(Story story)
    {
        Host.Instance.Application.Camera.TakeScreenshot();
    }
}
```

And set it up in your BDDfy config:

```
Configurator.Processors.Add(() => new SelenoExceptionProcessor());
```

**Comment by lchong0412** [comment]
@mwhelan Thank you, the extension method could be a solution to my problem. I also tried to implement the IProccessor, but the problem is I'm not using a static SelenoHost object, I'm using a per test class host, I got a problem trying to access the Host object in processor implementation.

**Comment by mwhelan** [comment]
We've designed SelenoHost to be a singleton because it's quite expensive to create. You don't want that overhead on every test. Is there a reason you want a new one per test? Perhaps it's to run different browsers. You can still follow the static SelenoHost approach and just have different instance properties for each browser - FireFoxInstance, ChromeInstance, etc. Would be interesting to see a test fixture....

**Comment by lchong0412** [comment]
@mwhelan Since xunit supports test class to run in parallel, I'm just creating a Host for each class so they can run in parallel. 
I create a host object and call run method for each test class (in the ClassFixture). Does it sound right?

**Comment by MehdiK** [comment]
@lchong0412 would you be able to send a bit of code so we know how you're using the frameworks? We might be able to help more that way.

**Comment by MehdiK** [comment]
I personally prefer the Processor model as suggested by @mwhelan.

**Comment by mwhelan** [comment]
@lchong0412 in that case I'd use a ConcurrentDictionary and have one SelenoHost per thread.

**Comment by lchong0412** [comment]
@mwhelan Yes, I ended up having a static class with a dictionary to keep all the SelenoHost object, it works for now. Thanks you all for your helps!

**Comment by baodinh77** [comment]
Hi,
How would I embed screenshot to the end of the scenario of html report when test fail? With Michael's example above, I change the code to save image to byte array, but I still don't know how to embed to the HTML report. Thanks for your helps!
